README.md
# NAME

Darkpan - A Mojolicious web service frontend leveraging OrePAN2

# DESCRIPTION

Darkpan is a webservice build on Mojolicious to frontend [OrePAN2](https://metacpan.org/pod/OrePAN2). This module was inspired
by [OrePAN2::Server](https://metacpan.org/pod/OrePAN2::Server&#x27;) but built on 
Mojolicious to take advantage of it's robust framework of tools. A good bit of the documentation
was also taken from OrePAN2::Server as the functionality is similar if not identical.

# SYNOPSIS

## Running the server

    # start a server with default configurations on port 8080
    darkpan --port 8080
    
    # start a server with AD backed basic auth
    # config.json
    {
      "basic_auth": {
        "Realm Name": {
          "host": "ldaps://my.ldap.server.org",
          "port": 636,
          "basedn": "DC=my,DC=compoany,DC=org",
          "binddn": "bind_name",
          "bindpw": "bond_pw",
          "filter": "(&(objectCategory=*)(sAMAccountName=%s)(|(objectClass=user)(objectClass=group)))"
        }
      }
    }
    
    darkpan --config ./config.json

#### Options:

- **-c,--config** _default: undef_: 
    JSON configuration file location
- **-p,--port** _default: 3000_: 
    Web application port

## Configuring the server

Configurations can be done using environment variables or by creating a json config file.

### environment variables

Environment variables are at a higher order than values set in the json configuration 
file and will take precedence if set. The **DARKPAN\_CONFIG\_FILE** env variable can be 
set instead of using the command line option to denote the path of the config file.

- **DARKPAN\_CONFIG\_FILE**: 
    Location of a json configuration file, same as passing the --config option
- **DARKPAN\_DIRECTORY**:
     Directory where uploads will be stored.
- **DARKPAN\_PATH**:
    URL path to function as cpan repository resolver/mirror endpoint.
- **DARKPAN\_COMPRESS\_INDEX**:
    Setting whether to compress (gzip) the 02packages.details.txt file.
- **DARKPAN\_AUTH\_REALM**:
    Basic auth realm name. This variable needs to be set to enable parsing of additional
    auth settings using the following the format DARKPAN\_AUTH\_\[setting\].
- **DARKPAN\_AUTH\_\[setting\]**:
    Additional basic auth settings can be passed using this format. The settings
    will be parsed and applied to your basic auth configurations. See
    [Mojolicious::Plugin::BasicAuthPlus](https://metacpan.org/pod/Mojolicious::Plugin::BasicAuthPlus) 
    for additional details on configurations.

### config.json

The config.json file contains customizations for the darkpan web application. It can be
referenced by absolute path or relative to where the application is run from.

    {
      "directory": "darkpan",
      "compress_index": true,
      "path": "darkpan,
      "basic_auth": {
        "Realm Name": {
          "host": "ldaps://my.ldap.server.org",
          "port": 636,
          "basedn": "DC=my,DC=compoany,DC=org",
          "binddn": "bind_name",
          "bindpw": "bond_pw",
          "filter": "(&(objectCategory=*)(sAMAccountName=%s)(|(objectClass=user)(objectClass=group)))"
        }
      }
    }

#### config.json options

- **directory** _default: darkpan_: 
    Directory where uploads will be stored.
- **path** _default: darkpan_: 
    URL path to function as cpan repository resolver/mirror endpoint.
- **compress\_index** _default: true_: 
    Setting whether to compress (gzip) the 02packages.details.txt file.
- **basic\_auth** _default: undef_: 
    Basic authentication settings, see configurations for [Mojolicious::Plugin::BasicAuthPlus](https://metacpan.org/pod/Mojolicious::Plugin::BasicAuthPlus). When not provided
    no authentication is necessary to post modules to the service.

### Authentication

Authentication is handled by [Mojolicious::Plugin::BasicAuthPlus](https://metacpan.org/pod/Mojolicious::Plugin::BasicAuthPlus).
To configure basic auth, see the configuration options for [Mojolicious::Plugin::BasicAuthPlus](https://metacpan.org/pod/Mojolicious::Plugin::BasicAuthPlus)
and add your settings to the basic auth section of the config.json file.     

## How to Deploy 

### Deploying with POST

Publishing to darkpan can be done using a post request and a URL to git or bitbucket repo.

    #upload git managed module to my darkpan by curl 
    curl --data-urlencode 'module=git@github.com:Songmu/p5-App-RunCron.git' --data-urlencode 'author=SHINGLER' http://localhost:3000/publish
    curl --data-urlencode 'module=git+ssh://git@mygit/home/git/repos/MyModule.git' --data-urlencode 'author=SHINGLER' http://localhost:3000/publish
    curl --data-urlencode 'module=git+file:///home/hirobanex/project/MyModule.git' --data-urlencode 'author=SHINGLER' http://localhost:3000/publish

### Deploying with [Minilla](https://metacpan.org/pod/Minilla)

Minilla is a cpan authoring tool, see [Minilla](https://metacpan.org/pod/Minilla) for more details.

#### minil.toml

Add a reference to a pause configruation file in your minil.toml that points to your darkpan instance. 
The configuration can reference a relative path as follows:

    [release]
    pause_config=".pause"

#### .pause file

The .pause file is a configuration file for uploading modules to CPAN or your own Darkpan Repository.
See [CPAN::Uploader](https://metacpan.org/pod/CPAN::Uploader) for more detail.

    upload_uri http://my-darkpan.server/publish
    user myUsername
    password myPassword
    

_\*\* if you don't set the upload\_uri, you will upload to CPAN_

If basic auth is enabled, the username and password set in the .pause file will be
used as basic auth credentials.

# LICENSE

Copyright (C) shingler.

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

# AUTHOR

shingler <shingler@oclc.org>


Changes
Revision history for Perl extension Mojo::Darkpan

{{$NEXT}}

0.09 2022-11-20T18:20:24Z
    - Fixed an error where bad regex was causing similarly named packages to show up in the "other_versions" section when looking at the /list API endpoint
    - Updated the included Dockerfile to use smaller Alpine based image. Noted some warnings when running in Alpine Docker, will investigate later
    - Added .dockerignore file to exclude some files/folders when doing Docker build

0.08 2022-09-15T14:47:49Z
    - Documentation fixes

0.07 2022-09-15T14:42:40Z
    - Updated documentation with additional example for publishing local files
    - Canged wording from "Deploy" to "Publish" in doco
    - Fixed example JSON config to add missing double quote in JSON definition

0.06 2021-12-13T18:02:19Z
    - Added missing depency
    - Removed use v5.25 in Mojo::Darkpan::Util since min version set to v5.20

0.05 2021-12-09T17:50:15Z
    - Changed how indexing happens when uploading. Found that OrePAN2 when recreating the index can take a lot of time when you have many packages in your repo. Complete reindexing will use a subprocess
    - Added /reindex to manually kick off index creation
    - Fixed issues in /list process where provides names were being added incorrectly to other packages in JSON response
    - Added convenience routines to Mojo::Darkpan::Config
    - Added routines to Mojo::Darkpan::Util for indexing

0.04 2021-12-08T16:02:53Z
    - removed some debug logging
    - fixed invalid reference in Config.pm

0.02 2021-12-08T02:05:28Z
    - style changes to directory handler plugin

0.01 2021-12-08T01:04:36Z

    - original version



lib/Darkpan/Util.pm
package Darkpan::Util;
use v5.25;
use Moo;
use Darkpan::Config;
use Data::Dumper;
use File::Temp;
use IO::Zlib;
use File::Spec;
use File::Basename;
use FindBin;
use Cwd;

has controller => (is => 'rw', required => 1);
has userinfo => (is => 'lazy');
has upload => (is => 'lazy');
has author => (is => 'lazy');
has indexer => (is => 'lazy');
has config => (is => 'lazy');


sub authorized {
    my $self = shift;
    if ($self->config->basic_auth) {

        my ($hash_ref, $auth_ok) = $self->controller->basic_auth(
            $self->config->basic_auth->{realm}
                => $self->config->basic_auth->{config}
        );

        if (!$auth_ok) {
            $self->controller->res->headers->www_authenticate('Basic');
            $self->controller->render(text => 'Authentication required!', status => 401);
            return undef;
        }
        else {
            return $auth_ok;
        }

    }
    else {
        return 1;
    }
}

sub publish {
    my $self = shift;
    my $module;
    my $author = $self->author;

    if ($self->upload) {
        # get a Mojo::Asset::File ref
        my $file = $self->upload->asset->to_file;

        # request from CPAN::Uploader
        my $tempdir = File::Temp::tempdir(CLEANUP => 1);

        $module = File::Spec->catfile($tempdir, $self->upload->filename);
        $file->move_to($module);
    }
    else {
        $module = $self->controller->param('module'); # can be a git repo.
    }

    if (!defined($module)) {
        die("no module provided in request")
    }

    my $injector = OrePAN2::Injector->new(
        directory => $self->config->directory,
        author    => $author,
    );

    $injector->inject($module);

    # reindex to create the modules file 
    $self->index();
}

sub index {
    my $self = shift;
    $self->indexer->make_index(
        no_compress => !$self->config->compressIndex,
    );
}

sub list {
    my $self = shift;
    my $no_compress = !$self->config->compressIndex;
    my $pkgfname = File::Spec->catfile(
        $self->config->directory,
        'modules',
        $no_compress ? '02packages.details.txt' : '02packages.details.txt.gz'
    );

    my $data;
    my $current;
    my $fh = IO::Zlib->new($pkgfname, "rb");
    while (<$fh>) {
        next if ($_ !~ m/\.tar\.gz$/);
        my ($name, $version, $file) = split('\s+', $_);
        if (eval($version)) {
            $current = $name;
            my $dir = File::Basename::dirname($file);
            my $archive = $file =~ s/$dir\///gr;
            $data->{$current}->{version} = $version;
            $data->{$current}->{archive} = $archive;
            $data->{$current}->{dir} = $dir;
            $data->{$current}->{other_versions} = $self->_getFileList($dir, $archive);
        }
        else {
            push(@{$data->{$current}->{provides}}, $name)
        }
    }

    return $data;
}

sub _getFileList {
    my $self = shift;
    my $path = shift;
    my $current = shift;

    my $base = getcwd;
    my $pkgdir = File::Spec->catfile(
        $self->config->directory,
        'authors',
        'id',
        $path
    );
    opendir my $dir, $pkgdir or die "Cannot open directory: $pkgdir";
    my @files = readdir $dir;
    closedir $dir;

    my @data;
    for (@files) {
        next if $_ =~ m/^\./;
        next if $_ eq $current;
        push(@data, $_);
    }

    return \@data;
}

sub _build_userinfo {
    my $self = shift;
    my ($username, $password) = split(':', $self->req->url->base->{userinfo});
    return {
        username => $username,
        password => $password
    }
}

sub _build_upload {
    my $self = shift;
    for my $file (@{$self->controller->req->uploads('files')}) {
        if ($file->name eq 'pause99_add_uri_httpupload') {
            return $file;
        }
    }
}

sub _build_author {
    my $self = shift;
    my $author = 'DUMMY';

    if ($self->controller->param("HIDDENNAME")) {
        $author = uc($self->controller->param("HIDDENNAME"));
    }
    elsif ($self->controller->param("author")) {
        $author = uc($self->request->param("author"));
    }

    return $author;
}

sub _build_indexer {
    my $self = shift;
    return OrePAN2::Indexer->new(directory => $self->config->directory);
}

sub _build_config {
    return Darkpan::Config->new;
}

1;

lib/Darkpan/Config.pm
package Darkpan::Config;
use v5.20;
use Moo;
use JSON;
use Mojo::File;
use Cwd;
use Data::Dumper;

has _config => (is => 'lazy');
has directory => (is => 'lazy');
has compressIndex => (is => 'lazy');
has path => (is => 'lazy');
has basic_auth => (is => 'lazy');

sub _build_path {
    my $self = shift;
    my $path;
    if ($ENV{DARKPAN_PATH}) {
        $path = $ENV{DARKPAN_PATH};
    }
    else {
        $path = $self->_config->{path}
            if $self->_config && $self->_config->{path};
    }

    # default if undef
    $path //= 'darkpan';

    return $path;
}

sub _build_basic_auth {
    my $self = shift;
    my $config;
    if ($ENV{DARKPAN_AUTH_REALM}) {
        $config->{realm} = $ENV{DARKPAN_AUTH_REALM};
        for my $k (keys %ENV) {
            if ($k =~ m/^DARKPAN_AUTH_(.*)/) {
                $config->{config}->{lc($1)} = $ENV{$k};
            }
        }
    }
    elsif (defined($self->_config) && defined($self->_config->{basic_auth})) {
        my @keys = keys %{$self->_config->{basic_auth}};
        $config = {
            realm  => $keys[0],
            config => $self->_config->{basic_auth}->{$keys[0]}
        };
    }
    
    return $config;
}

sub _build__config {
    my $self = shift;
    my $location = $ENV{DARKPAN_CONFIG_FILE};

    if (defined($location)) {

        $self->_validateAssetLocation($location);

        my $file = Mojo::File->new($location);
        my $config = JSON->new->utf8->decode($file->slurp);

        return $config;
    }
    
    return undef;
}

sub _build_directory {
    my $self = shift;

    my $dir;
    if ($ENV{DARKPAN_DIRECTORY}) {
        $dir = $ENV{DARKPAN_DIRECTORY};
    }
    else {
        $dir = $self->_config->{directory}
            if $self->_config && $self->_config->{directory};
    }

    # default if undef
    $dir //= 'darkpan';

    $self->_validateAssetLocation($dir);

    return $dir;
}

sub _build_compressIndex {
    my $self = shift;
    my $compress;
    if ($ENV{DARKPAN_COMPRESS_INDEX}) {
        $compress = $ENV{DARKPAN_COMPRESS_INDEX};
    }
    else {
        $compress = $self->_config->{compress_index}
            if $self->_config && $self->_config->{compress_index};
    }

    # default if undef
    $compress //= 1;

    return $compress;
}

sub _validateAssetLocation {
    my $dir = shift;

    if ($dir !~ m/^\//) {
        $dir =~ s/^\.\///;
        my $base = getcwd;
        $dir = "$base/$dir";
    }

    return $dir;
}

1;

lib/Mojolicious/Plugin/DirectoryHandler.pm
package Mojolicious::Plugin::DirectoryHandler;
use Mojo::Base qw{Mojolicious::Plugin};
use Cwd ();
use Encode ();
use DirHandle;
use Mojolicious::Types;
use Mojo::JSON qw(encode_json);
use Darkpan::Config;

my $dir_page = <<'PAGE';
<html>
<head>
    <title>Index of <%= $cur_path %></title>
    <meta http-equiv="content-type" content="text/html; charset=utf-8" />
    <style>

        table { width:100%; }

        td, th{ padding: 0 5px 0 5px }

        .name { text-align:left; }

        .size, .mtime { text-align:right; }

        .type { width:11em;text-align:left; }

        .mtime { width:15em; }

    </style>
</head>
<body>
<h1>Index of <%= $cur_path %></h1>
<hr />
<table>
    <tr>
        <th class='name'>Name</th>
        <th class='size'>Size</th>
        <th class='type'>Type</th>
        <th class='mtime'>Last Modified</th>
    </tr>
    % for my $file (@$files) {
    <tr>
        <td class='name'><a href='<%= $file->{url} %>'><%== $file->{name} %></a></td>
        <td class='size'><%= $file->{size} %></td>
        <td class='type'><%= $file->{type} %></td>
        <td class='mtime'><%= $file->{mtime} %></td></tr>
    % }
</table>
<hr />
</body></html>
PAGE

my $types = Mojolicious::Types->new;

sub register {
    my $self = shift;
    my ($app, $args) = @_;

    my $root = Mojo::Home->new($args->{root} || Darkpan::Config->new->directory);
    my $handler = $args->{handler};
    my $auto_index = $args->{auto_index} // 1;
    my $json = $args->{json};
    my $delivery_path = $args->{delivery_path} || $root;
    $dir_page = $args->{dir_page} if ($args->{dir_page});

    $app->hook(
        before_dispatch => sub {
            my $c = shift;
            my $req_path = $c->req->url->path;

            return render_file($c, $root, $handler) if (-f $root->to_string());

            my $path;

            if ($req_path =~ m/\/$delivery_path/) {
                my $root_string = $root->to_string();
                $req_path =~ s/$delivery_path//;
                if ($req_path ne $root_string) {
                    $path = $root->rel_file(Mojo::Util::url_unescape($req_path));
                }

                if (-d $path) {
                    render_indexes($c, $path, $json) unless not $auto_index;
                    return;
                }
                if (-f $path) {
                    render_file($c, $path, $handler);
                    return;
                }
                $c->redirect_to($c->req->url->path->trailing_slash(1));
            }
        },
    );
    return $app;
}

sub render_file {
    my $c = shift;
    my $path = shift;
    my $handler = shift;
    $handler->($c, $path) if (ref $handler eq 'CODE');
    return if ($c->tx->res->code);
    my $data = Mojo::File::slurp($path);
    $c->render(data => $data, format => get_ext($path) || 'txt');
}

sub render_indexes {
    my $c = shift;
    my $dir = shift;
    my $json = shift;

    my @files = ($c->req->url->path eq '/') ? ()
        : ({ url => '../', name => 'Parent Directory', size => '', type => '', mtime => '' });
    my $children = list_files($dir);

    my $cur_path = Encode::decode_utf8(Mojo::Util::url_unescape($c->req->url->path));
    for my $basename (sort {$a cmp $b} @$children) {
        my $file = "$dir/$basename";
        my $url = Mojo::Path->new($cur_path)->trailing_slash(0);
        push @{$url->parts}, $basename;

        my $is_dir = -d $file;
        my @stat = stat _;
        if ($is_dir) {
            $basename .= '/';
            $url->trailing_slash(1);
        }

        my $mime_type = $is_dir ? 'directory'
            : ($types->type(get_ext($file) || 'txt') || 'text/plain');
        my $mtime = Mojo::Date->new($stat[9])->to_string();

        push @files, {
            url   => $url,
            name  => $basename,
            size  => $stat[7] || 0,
            type  => $mime_type,
            mtime => $mtime,
        };
    }

    my $any = { inline => $dir_page, files => \@files, cur_path => $cur_path };
    if ($json) {
        $c->respond_to(
            json => { json => encode_json(\@files) },
            any  => $any,
        );
    }
    else {
        $c->render(%$any);
    }
}

sub get_ext {
    $_[0] =~ /\.([0-9a-zA-Z]+)$/ || return;
    return lc $1;
}

sub list_files {
    my $dir = shift || return [];
    my $dh = DirHandle->new($dir);
    my @children;
    while (defined(my $ent = $dh->read)) {
        next if $ent eq '.' or $ent eq '..';
        push @children, Encode::decode_utf8($ent);
    }
    return [ @children ];
}

1;

__END__

lib/Mojo/Darkpan.pm
package Mojo::Darkpan;
use v5.20;
use Data::Dumper;
use FindBin;
use Mojo::Base 'Mojolicious', -signatures;
use Mojo::Darkpan::Config;

sub startup($self) {
    # https://metacpan.org/pod/Mojolicious::Plugin::BasicAuthPlus
    $self->plugin('basic_auth_plus');
    $self->plugin('DirectoryHandler',
        delivery_path => Mojo::Darkpan::Config->new->path);

    #----------
    # Router
    #----------
    my $r = $self->routes;
    my $index = $r->any('/');
    $index->get('/list')->to(controller => 'index', action => 'list');
    $index->get('/reindex')->to(controller => 'index', action => 'reindex');

    my $uploader = $r->any('/publish');
    $uploader->post('/')->to(controller => 'publish', action => 'upload');

    # goes to the same place as publish, used for compatibility
    my $authenquery = $r->any('/authenquery');
    $authenquery->post('/')->to(controller => 'publish', action => 'upload');

}

our $VERSION = "0.07";

1;
__END__

=encoding utf-8

=head1 NAME

Mojo::Darkpan - A Mojolicious web service frontend leveraging OrePAN2

=head1 DESCRIPTION

Mojo::Darkpan is a webservice build on Mojolicious to frontend L<OrePAN2|https://metacpan.org/pod/OrePAN2>. This module was inspired
by L<OrePAN2::Server|https://metacpan.org/pod/OrePAN2::Server'> but built on 
Mojolicious to take advantage of it's robust framework of tools. A good bit of the documentation
was also taken from OrePAN2::Server as the functionality is similar if not identical.

=head1 SYNOPSIS

=head2 Running the server

    # start a server with default configurations on port 8080
    darkpan --port 8080
    
    # start a server with AD backed basic auth
    # config.json
    {
      "basic_auth": {
        "Realm Name": {
          "host": "ldaps://my.ldap.server.org",
          "port": 636,
          "basedn": "DC=my,DC=compoany,DC=org",
          "binddn": "bind_name",
          "bindpw": "bond_pw",
          "filter": "(&(objectCategory=*)(sAMAccountName=%s)(|(objectClass=user)(objectClass=group)))"
        }
      }
    }
    
    darkpan --config ./config.json

=head4 Options:

=over 2

=item B<-c,--config> I<default: undef>: 
    JSON configuration file location

=item B<-p,--port> I<default: 3000>: 
    Web application port

=back 

=head4 Paths:

=over 2

=item B</publish> or B</authenquery>:
    POST path(s) for releasing packages, set as the upload_uri in your .pause file

=item B</list>: 
    JSON list of packages and info about them
    
=item B</reindex>: 
    Force OrePAN2 to do a index all modules and recreate the 02packages.details file. 
    
=item B</darkpan> I<(may differ if you set a custom path)>: 
    Directory listing of the repository. This is the path to reference as your mirror or
    set the PERL_CARTON_MIRROR env var       

=back  

=head2 Configuring the server

Configurations can be done using environment variables or by creating a json config file.

=head3 environment variables

Environment variables are at a higher order than values set in the json configuration 
file and will take precedence if set. The B<DARKPAN_CONFIG_FILE> env variable can be 
set instead of using the command line option to denote the path of the config file.

=over 2

=item B<DARKPAN_CONFIG_FILE>: 
    Location of a json configuration file, same as passing the --config option

=item B<DARKPAN_DIRECTORY>:
     Directory where uploads will be stored.

=item B<DARKPAN_PATH>:
    URL path to function as cpan repository resolver/mirror endpoint.
    
=item B<DARKPAN_COMPRESS_INDEX>:
    Setting whether to compress (gzip) the 02packages.details.txt file.

=item B<DARKPAN_AUTH_REALM>:
    Basic auth realm name. This variable needs to be set to enable parsing of additional
    auth settings using the following the format DARKPAN_AUTH_[setting].
    
=item B<DARKPAN_AUTH_[setting]>:
    Additional basic auth settings can be passed using this format. The settings
    will be parsed and applied to your basic auth configurations. See
    L<Mojolicious::Plugin::BasicAuthPlus|https://metacpan.org/pod/Mojolicious::Plugin::BasicAuthPlus> 
    for additional details on configurations.

=back

=head3 config.json

The config.json file contains customizations for the darkpan web application. It can be
referenced by absolute path or relative to where the application is run from.

    {
      "directory": "darkpan",
      "compress_index": true,
      "path": "darkpan,
      "basic_auth": {
        "Realm Name": {
          "host": "ldaps://my.ldap.server.org",
          "port": 636,
          "basedn": "DC=my,DC=compoany,DC=org",
          "binddn": "bind_name",
          "bindpw": "bond_pw",
          "filter": "(&(objectCategory=*)(sAMAccountName=%s)(|(objectClass=user)(objectClass=group)))"
        }
      }
    }

=head4 config.json options

=over 2

=item B<directory> I<default: darkpan>: 
    Directory where uploads will be stored.

=item B<path> I<default: darkpan>: 
    URL path to function as cpan repository resolver/mirror endpoint.

=item B<compress_index> I<default: true>: 
    Setting whether to compress (gzip) the 02packages.details.txt file.

=item B<basic_auth> I<default: undef>: 
    Basic authentication settings, see configurations for L<Mojolicious::Plugin::BasicAuthPlus|https://metacpan.org/pod/Mojolicious::Plugin::BasicAuthPlus>. When not provided
    no authentication is necessary to post modules to the service.

=back

=head3 Authentication

Authentication is handled by L<Mojolicious::Plugin::BasicAuthPlus|https://metacpan.org/pod/Mojolicious::Plugin::BasicAuthPlus>.
To configure basic auth, see the configuration options for L<Mojolicious::Plugin::BasicAuthPlus|https://metacpan.org/pod/Mojolicious::Plugin::BasicAuthPlus>
and add your settings to the basic auth section of the config.json file.     

=head2 How to Deploy 

=head3 Deploying with POST

Publishing to darkpan can be done using a post request and a URL to git or bitbucket repo.
    
    #upload git managed module to my darkpan by curl 
    curl --data-urlencode 'module=git+ssh://git@mygit/home/git/repos/MyModule.git' --data-urlencode 'author=reshingleton' http://localhost:3000/publish
    curl --data-urlencode 'module=git+file:///home/rshingleton/project/MyModule.git' --data-urlencode 'author=reshingleton' http://localhost:3000/publish
    curl --data-urlencode 'module=git@github.com:rshingleton/perl-module-test.git' --data-urlencode 'author=reshingleton' http://localhost:3000/publish

The module parameter can also be an HTTP url. see L<OrePAN2::Injector|https://metacpan.org/pod/OrePAN2::Injector> for 
additional details.

    curl --data-urlencode 'module=https://cpan.metacpan.org/authors/id/O/OA/OALDERS/OrePAN2-0.48.tar.gz' --data-urlencode 'author=OALDERS' http://localhost:3000/publish
    
=head3 Deploying with L<Minilla|https://metacpan.org/pod/Minilla>

Minilla is a cpan authoring tool, see L<Minilla|https://metacpan.org/pod/Minilla> for more details.

=head4 minil.toml

Add a reference to a pause configruation file in your minil.toml that points to your darkpan instance. 
The configuration can reference a relative path as follows:
    
    [release]
    pause_config=".pause"

=head4 .pause file

The .pause file is a configuration file for uploading modules to CPAN or your own Darkpan Repository.
See L<CPAN::Uploader|https://metacpan.org/pod/CPAN::Uploader> for more detail.

    upload_uri http://my-darkpan.server/publish
    user myUsername
    password myPassword
    
I<** if you don't set the upload_uri, you will upload to CPAN>

If basic auth is enabled, the username and password set in the .pause file will be
used as basic auth credentials.

=head2 How to install from your Darkpan

=head3 cpanm

See L<cpanm|https://metacpan.org/pod/cpanm> for additional details.

     # check CPAN and your Darkpan server
     cpanm --mirror http://my-darkpan.server/darkpan
     
     # check for packages from only your Darkpan server
     cpanm --mirror-only http://my-darkpan.server/darkpan
     cpanm --from http://my-darkpan.server/darkpan

=head3 cpm

See L<cpm|https://metacpan.org/dist/App-cpm/view/script/cpm> for additional details.

    # resolve distribution names from DARKPAN/modules/02packages.details.txt.gz
    # and fetch distibutions from DARKPAN/authors/id/...
    > cpm install --resolver 02packages,http://example.com/darkpan Your::Module
    
    # use darkpan first, and if it fails, use metadb and normal CPAN
    > cpm install --resolver 02packages,http://my-darkpan.server/darkpan --resolver metadb Your::Module

=head3 carton

See L<carton|https://metacpan.org/pod/Carton> for additional details.
 
    # in the cpanfile
    # local mirror (darkpan)
    
    requires 'Plack', '== 0.9981',
      dist => 'MYCOMPANY/Plack-0.9981-p1.tar.gz',
      mirror => 'http://my-darkpan.server/darkpan';

Carton also uses an (L<undocumented|https://domm.plix.at/perl/2017_07_carton_darkpan.html>) environment variable PERL_CARTON_MIRROR that will enable you
to add your Darkpan server to its list of resolvers. Carton will install from 
your Darkpan and from the default CPAN mirror.   

    PERL_CARTON_MIRROR=http://my-darkpan.server/darkpan carton install  

=head1 SEE ALSO

L<OrePAN2|https://metacpan.org/pod/OrePAN2>

L<OrePAN2::Server|https://metacpan.org/pod/OrePAN2::Server>

 
=head1 LICENSE

Copyright (C) rshingleton.

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

=head1 AUTHOR

rshingleton E<lt>reshingleton@gmail.comE<gt>

=cut

